File: kex-runner/src/main/kotlin/org/jetbrains/research/kex/jacoco/TestsCompiler.java
package org.jetbrains.research.kex.jacoco;

import javax.tools.*;
import java.io.*;
import java.net.URI;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class TestsCompiler {

    private final List<ClassJavaFileObject> generatedClasses = new ArrayList<>();

    private List<ClassJavaFileObject> getGeneratedClasses(File javaFile) {

        JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();

        SimpleJavaFileManager fileManager =
                new SimpleJavaFileManager(compiler.getStandardFileManager(null, null, null));

        JavaFileObject compilationUnit = new TestJavaFileObject(javaFile);

        JavaCompiler.CompilationTask compilationTask = compiler.getTask(
                null, fileManager, null, null, null, Collections.singletonList(compilationUnit));

        compilationTask.call();

        return fileManager.getGeneratedOutputFiles();
    }

    public void generateAll(String directoryPath) {
        File directory = new File(directoryPath);
        if (!directory.isDirectory()) throw new IllegalArgumentException();
        File[] files = directory.listFiles();
        Objects.requireNonNull(files);
        for (File file : files) {
            if (!file.isDirectory()) {
                generatedClasses.addAll(getGeneratedClasses(file));
            }
            else {
                generateAll(file.getPath());
            }
        }
    }

    public List<String> getTestsNames() {
        List<String> result = new ArrayList<>();
        generatedClasses.forEach(c -> result.add(c.getClassName()));
        return result;
    }

    public CompiledClassLoader getCompiledClassLoader(URLClassLoader urlClassLoader) {
        return new CompiledClassLoader(generatedClasses, urlClassLoader);
    }

    public static class TestJavaFileObject extends SimpleJavaFileObject {

        private final File javaFile;

        public TestJavaFileObject(File javaFile) {
            super(URI.create(javaFile.getName()), Kind.SOURCE);
            this.javaFile = javaFile;
        }

        @Override
        public CharSequence getCharContent(boolean ignoreEncodingErrors) throws IOException {
            List<String> lines = Files.readAllLines(javaFile.toPath());
            StringBuilder sb = new StringBuilder();
            for (String line : lines) {
                sb.append(line).append('\n');
            }
            return sb.toString();
        }
    }

    private static class ClassJavaFileObject extends SimpleJavaFileObject {

        private final ByteArrayOutputStream outputStream;

        private final String className;

        protected ClassJavaFileObject(String className) {
            super(URI.create(className.replace('.', '/') + ".class"), Kind.CLASS);
            this.className = className;
            outputStream = new ByteArrayOutputStream();
        }

        @Override
        public OutputStream openOutputStream() {
            return outputStream;
        }

        public byte[] getBytes() {
            return outputStream.toByteArray();
        }

        public String getClassName() {
            return className;
        }
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    private static class SimpleJavaFileManager extends ForwardingJavaFileManager {

        private final List<ClassJavaFileObject> outputFiles;

        protected SimpleJavaFileManager(JavaFileManager fileManager) {
            super(fileManager);
            outputFiles = new ArrayList<>();
        }

        @Override
        public JavaFileObject getJavaFileForOutput(
                Location location, String className, JavaFileObject.Kind kind, FileObject sibling
        ) {
            ClassJavaFileObject file = new ClassJavaFileObject(className);
            outputFiles.add(file);
            return file;
        }

        public List<ClassJavaFileObject> getGeneratedOutputFiles() {
            return outputFiles;
        }
    }

    public static class CompiledClassLoader extends ClassLoader {

        private final List<ClassJavaFileObject> files;

        private final URLClassLoader parent;

        private CompiledClassLoader(List<ClassJavaFileObject> files, URLClassLoader parent) {
            this.files = files;
            this.parent = parent;
        }

        @Override
        public Class<?> loadClass(String name) throws ClassNotFoundException {
            byte[] bytes = getBytes(name);
            if (bytes != null) {
                return defineClass(name, bytes, 0, bytes.length);
            }
            return parent.loadClass(name);
        }

        public byte[] getBytes(String name) {
            for (ClassJavaFileObject file : files) {
                if (file.getClassName().equals(name)) {
                    return file.getBytes();
                }
            }
            return null;
        }

        @Override
        public InputStream getResourceAsStream(String name) {
            return parent.getResourceAsStream(name);
        }

        public URL[] getURLs() {
            return parent.getURLs();
        }
    }

}

File: kex-runner/src/main/kotlin/org/jetbrains/research/kex/jacoco/TestsCompiler.kt
package org.jetbrains.research.kex.jacoco

import org.jetbrains.research.kthelper.tryOrNull
import java.io.*
import java.net.URI
import java.net.URL
import java.net.URLClassLoader
import java.nio.file.Files
import java.nio.file.Path
import java.util.stream.Collectors
import javax.tools.*
import kotlin.io.path.isRegularFile

class TestsCompiler(private val outputDir: Path) {
    private val generatedClasses: MutableList<ClassJavaFileObject> = ArrayList()

    private fun getGeneratedClasses(javaFile: File): List<ClassJavaFileObject> {
        val compiler = ToolProvider.getSystemJavaCompiler()

        val standardFileManager = compiler.getStandardFileManager(null, null, null)
        standardFileManager.setLocation(StandardLocation.CLASS_OUTPUT, listOf(outputDir.toFile()))
        val fileManager = SimpleJavaFileManager(
            standardFileManager
        )
        val compilationUnit: JavaFileObject = TestJavaFileObject(javaFile)
        val compilerOutput = ByteArrayOutputStream()
        val compilationTask = compiler.getTask(
            compilerOutput.writer(), fileManager, null, null, null, listOf(compilationUnit)
        )
        compilationTask.call()
        return fileManager.generatedOutputFiles
    }

    fun generateAll(directoryPath: Path) = tryOrNull {
        Files.walk(directoryPath)
            .filter { it.isRegularFile() }
            .map { it.toFile() }
            .map { getGeneratedClasses(it) }
            .forEach { generatedClasses.addAll(it) }
    }

    val testsNames: List<String>
        get() = generatedClasses
            .stream()
            .map { obj: ClassJavaFileObject -> obj.className }
            .collect(Collectors.toList())

    fun getCompiledClassLoader(urlClassLoader: URLClassLoader): CompiledClassLoader {
        return CompiledClassLoader(generatedClasses, urlClassLoader)
    }

    class TestJavaFileObject(
        private val javaFile: File
    ) : SimpleJavaFileObject(
        URI.create(javaFile.name),
        JavaFileObject.Kind.SOURCE
    ) {
        override fun getCharContent(ignoreEncodingErrors: Boolean): CharSequence =
            Files.readAllLines(javaFile.toPath()).joinToString("\n")
    }

    class ClassJavaFileObject(
        val className: String
    ) : SimpleJavaFileObject(
        URI.create(className.replace('.', '/') + ".class"),
        JavaFileObject.Kind.CLASS
    ) {
        private val outputStream = ByteArrayOutputStream()

        val bytes: ByteArray
            get() = outputStream.toByteArray()

        override fun openOutputStream(): OutputStream = outputStream
    }

    private class SimpleJavaFileManager(
        fileManager: JavaFileManager
    ) : ForwardingJavaFileManager<JavaFileManager>(fileManager) {
        private val outputFiles = mutableListOf<ClassJavaFileObject>()

        val generatedOutputFiles: List<ClassJavaFileObject>
            get() = outputFiles

        override fun getJavaFileForOutput(
            location: JavaFileManager.Location,
            className: String,
            kind: JavaFileObject.Kind,
            sibling: FileObject
        ): JavaFileObject = ClassJavaFileObject(className).also { outputFiles.add(it) }
    }

    class CompiledClassLoader(
        val files: List<ClassJavaFileObject>,
        val parent: URLClassLoader
    ) : ClassLoader() {
        val urLs: Array<URL>
            get() = parent.urLs

        override fun loadClass(name: String): Class<*> {
            val bytes = getBytes(name)
            return if (bytes != null) {
                defineClass(name, bytes, 0, bytes.size)
            } else parent.loadClass(name)
        }

        fun getBytes(name: String): ByteArray? {
            for (file in files) {
                if (file.className == name) {
                    return file.bytes
                }
            }
            return null
        }

        override fun getResourceAsStream(name: String): InputStream {
            return parent.getResourceAsStream(name)!!
        }

    }
}